### PR TITLE
[SPARK-49017][SQL] Insert statement fails when multiple parameters are being used

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Fix for multiple parameters support.

### Why are the changes needed?
The use of multiple parameters with identifiers were broken

### Does this PR introduce _any_ user-facing change?
Yes, look at tests.

### How was this patch tested?
New tests added to ParametersSuite.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by mihailom-db** [comment]
@cloud-fan could you take a look at this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 1]
```diff
@@ -60,9 +60,25 @@ trait UnresolvedUnaryNode extends UnaryNode with UnresolvedNode
  */
 case class PlanWithUnresolvedIdentifier(
     identifierExpr: Expression,
-    planBuilder: Seq[String] => LogicalPlan)
-  extends UnresolvedLeafNode {
+    otherPlans: Seq[LogicalPlan],
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:63]: maybe just call it `children`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala [hunk 2]
```diff
@@ -177,9 +177,17 @@ object BindParameters extends ParameterizedQueryProcessor with QueryErrorsBase {
   }
 
   private def bind(p: LogicalPlan)(f: PartialFunction[Expression, Expression]): LogicalPlan = {
-    p.resolveExpressionsWithPruning(_.containsPattern(PARAMETER)) (f orElse {
-      case sub: SubqueryExpression => sub.withNewPlan(bind(sub.plan)(f))
-    })
+    p match {
+      case lp: PlanWithUnresolvedIdentifier =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:181]: Do we need to do it? `resolveExpressionsWithPruning` already recursively transforms all the tree nodes and handles their expressions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIdentifierClause.scala [hunk 3]
```diff
@@ -37,7 +37,7 @@ class ResolveIdentifierClause(earlyBatches: Seq[RuleExecutor[LogicalPlan]#Batch]
   override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUpWithPruning(
     _.containsAnyPattern(UNRESOLVED_IDENTIFIER, UNRESOLVED_IDENTIFIER_WITH_CTE)) {
     case p: PlanWithUnresolvedIdentifier if p.identifierExpr.resolved =>
-      executor.execute(p.planBuilder.apply(evalIdentifierExpr(p.identifierExpr)))
+      executor.execute(p.planBuilder.apply(evalIdentifierExpr(p.identifierExpr), p.otherPlans))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIdentifierClause.scala:40]: I think a problem is when to invoke the plan builder. Shall we wait for all `p.otherPlans` to be resolved first? Without a clear trigger condition, we may hit rule order issues.

Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIdentifierClause.scala:40]: Yeah, that is true. That is why I tried to add a couple of tests with interleaving of identifiers and non-identifiers. I can pass one more time through AstBuilder, as I believe only InsertInto and one more LogicalPlans were changed to have otherPlans as parameters.

File: sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala [hunk 4]
```diff
@@ -624,6 +624,67 @@ class ParametersSuite extends QueryTest with SharedSparkSession with PlanTest {
...
+      // Select from table using param - WORKS
+      checkAnswer(spark.sql("select * from identifier(:tab)", Map("tab" -> "testtab")),
+        Seq(Row(1, "test1")))
+
+      // Insert into table using multiple params - WORKS
+      spark.sql("insert into identifier(:tab) values(2, :name)",
+        Map("tab" -> "testtab", "name" -> "test2"))
+
+      // Select from table using param - WORKS
+      checkAnswer(spark.sql("select * from identifier(:tab)", Map("tab" -> "testtab")),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala:644]: This is a duplicated test as https://github.com/apache/spark/pull/47501/files#diff-509971ca229888a8895c9402f41a4ed82989e2678fff4036e9b6b03dfa253fdaR636

File: sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala [hunk 5]
```diff
@@ -624,6 +624,67 @@ class ParametersSuite extends QueryTest with SharedSparkSession with PlanTest {
...
+        Seq(Row(1, "test1"), Row(2, "test2")))
+
+      // Insert into table using multiple params and idents - WORKS
+      spark.sql("insert into identifier(:tab) values(2, :name)",
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala:648]: ditto, duplicated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 6]
```diff
@@ -68,23 +68,25 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper
 
   protected def withIdentClause(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:69]: can we create an overload of this method that takes no `otherPlans`? Then we can reduce a lot of code changes of this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala (hunk 4 above)
Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala:644]: Yeah, it is not necessary to do parameters here. Will change it to normal select, as the aim of this query was to check contents of the table after previous addition.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 7]
```diff
@@ -2565,7 +2580,7 @@ class AstBuilder extends DataTypeAstBuilder
     } else {
       // It's a function call
       val funcCtx = ctx.functionName
-      val func = withFuncIdentClause(
+      val func: Expression = withFuncIdentClause(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2583]: unnecessary change?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 7 above)
Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2583]: Nope, there are two withFuncIdentClause, one that takes LogicalPlan and other with Expression. Compiler was not smart to pick up which one it was.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 8]
```diff
@@ -288,13 +288,7 @@ object CTESubstitution extends Rule[LogicalPlan] {
...
         }.getOrElse(u)
 
-      case p: PlanWithUnresolvedIdentifier =>
-        // We must look up CTE relations first when resolving `UnresolvedRelation`s,
-        // but we can't do it here as `PlanWithUnresolvedIdentifier` is a leaf node
-        // and may produce `UnresolvedRelation` later.
-        // Here we wrap it with `UnresolvedWithCTERelations` so that we can
-        // delay the CTE relations lookup after `PlanWithUnresolvedIdentifier` is resolved.
-        UnresolvedWithCTERelations(p, cteRelations)
+      case p: PlanWithUnresolvedIdentifier => p // Just delay until bindings are done
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:291]: do we really need to delay it? It seems no harm to do the CTE relation resolution right now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 9]
```diff
@@ -574,7 +574,7 @@ class SparkSqlAstBuilder extends AstBuilder {
         throw QueryParsingErrors.defineTempViewWithIfNotExistsError(ctx)
       }
 
-      withIdentClause(ctx.identifierReference(), ident => {
+      withIdentClause(ctx.identifierReference(), Seq(qPlan), (ident, otherPlans) => {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:577]:
```suggestion
      withIdentClause(ctx.identifierReference(), Seq(plan), (ident, otherPlans) => {
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@mihailom-db can you fix the merge conflicts?

**Comment by mihailom-db** [comment]
@cloud-fan resolved conflicts, ready for merge

**Comment by mihailom-db** [comment]
@MaxGekk could you take a look?

File: sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala [hunk 10]
```diff
@@ -625,6 +625,63 @@ class ParametersSuite extends QueryTest with SharedSparkSession with PlanTest {
     comparePlans(expected, parameterizedSql)
   }
 
+  test("SPARK-49017: bind named parameters with IDENTIFIER clause") {
+    withTable("testtab") {
+      // Create table
+      spark.sql("create table testtab (id int, name string)")
+
+      // Insert into table using single param - WORKS
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala:633]: nit: The `WORKS` word is not informative. Please, remove it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 11]
```diff
@@ -453,12 +467,12 @@ class AstBuilder extends DataTypeAstBuilder
...
         val (relationCtx, options, cols, partition, ifPartitionNotExists, byName)
         = visitInsertIntoTable(table)
-        withIdentClause(relationCtx, ident => {
+        withIdentClause(relationCtx, Seq(query), (ident, otherPlans) => {
           val insertIntoStatement = InsertIntoStatement(
             createUnresolvedRelation(relationCtx, ident, options),
             partition,
             cols,
-            query,
+            otherPlans.head,
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:475]: Are you sure that everywhere you call `.head`, it won't fail with `NoSuchElementException`. Could you clarify your assumption, please.

Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:475]: For all the places where we call `.head` we have otherPlans that was defined as `Seq(query)`, so it will have exactly one element. But Seq was used, if in the future the need for multiple plans is needed.

**Review by MaxGekk — APPROVED** [review]
The last changes are trivial. Just waiting for a success build.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @mihailom-db and @cloud-fan for review.

**Comment by MaxGekk** [comment]
@mihailom-db You pointed out only 4.0.0 as the affected version in SPARK-49017. Do the previous versions suffer from the same issue, and shall we backport the changes to other branches. Please, double check and confirm.

**Comment by mihailom-db** [comment]
@MaxGekk checked 3.4 does not have option of parameter markers in `IDENTIFIER(...)`, but 3.5 seems to suffer from the same error, so backport is needed

**Comment by mihailom-db** [comment]
@MaxGekk I have checked now, and there are multiple fixes that were not backported to `branch-3.5`, so I would say this one does not have to be backported as well. For example a fix for `ExpressionWithUnresolvedIdentifier` was not backported and this blocks backport of this PR as well.

**Comment by MaxGekk** [comment]
> I have checked now, and there are multiple fixes that were not backported to branch-3.5

@mihailom-db Please, open separate PRs for the `branch-3.5` with bug fixes.
